src/combinatorics/simple_graph/subgraph.lean

```diff
@@ -0,0 +1,302 @@
+  inf_le_left := λ x y, ⟨set.inter_subset_left x.verts y.verts, (λ v w h, h.1)⟩,
+  inf_le_right := λ x y, ⟨set.inter_subset_right x.verts y.verts, (λ v w h, h.2)⟩ }
+
+/-- The top of the `subgraph G` lattice is equivalent to the graph itself. -/
+def top_equiv : (⊤ : subgraph G).coe ≃g G :=
+{ to_fun := λ v, ↑v,
+  inv_fun := λ v, ⟨v, by tidy⟩,
+  left_inv := by tidy,
+  right_inv := by tidy,
+  map_rel_iff' := by tidy }
+
+/-- The bottom of the `subgraph G` lattice is equivalent to the empty graph on the empty
+vertex type. -/
+def bot_equiv : (⊥ : subgraph G).coe ≃g empty_graph empty :=
+{ to_fun := λ v, false.elim v.property,
+  inv_fun := λ v, begin cases v, end,
+  left_inv := by tidy,
+  right_inv := by tidy,
+  map_rel_iff' := by tidy }
+
+/-- Given two subgraphs, one a subgraph of the other, there is an induced injective homomorphism of
+the subgraphs as graphs. -/
+def map {x y : subgraph G} (h : x ≤ y) : x.coe →g y.coe :=
+{ to_fun := λ v, ⟨↑v, and.left h v.property⟩,
+  map_rel' := λ v w hvw, h.2 hvw }
+
+lemma map.injective {x y : subgraph G} (h : x ≤ y) : function.injective (map h) :=
+λ v w h, by { simp only [map, rel_hom.coe_fn_mk, subtype.mk_eq_mk] at h, exact subtype.ext h }
+
+/-- There is an induced injective homomorphism of a subgraph of `G` into `G`. -/
+def map_top (x : subgraph G) : x.coe →g G :=
+{ to_fun := λ v, v,
+  map_rel' := λ v w hvw, x.adj_sub hvw }
+
+lemma map_top.injective {x : subgraph G} : function.injective x.map_top :=
+λ v w h, subtype.ext h
+
+@[simp]
+lemma map_top_to_fun {x : subgraph G} (v : x.verts) : x.map_top v = v := rfl
+
+lemma neighbor_set_subset_of_subgraph {x y : subgraph G} (h : x ≤ y) (v : V) :
+  x.neighbor_set v ⊆ y.neighbor_set v :=
+λ w h', h.2 h'
+
+instance neighbor_set.decidable_pred (G' : subgraph G) [h : decidable_rel G'.adj] (v : V) :
+  decidable_pred (G'.neighbor_set v) := h v
+
+/-- If a graph is locally finite at a vertex, then so is a subgraph of that graph. -/
+instance finite_at {G' : subgraph G} (v : G'.verts) [decidable_rel G'.adj]
+   [fintype (G.neighbor_set v)] : fintype (G'.neighbor_set v) :=
+set.fintype_subset (G.neighbor_set v) (G'.neighbor_set_subset v)
+
+/-- If a subgraph is locally finite at a vertex, then so are subgraphs of that subgraph.
+
+This is not an instance because `G''` cannot be inferred. -/
+def finite_at_subgraph {G' G'' : subgraph G} [decidable_rel G'.adj]
+   (h : G' ≤ G'') (v : G'.verts) [hf : fintype (G''.neighbor_set v)] :
+   fintype (G'.neighbor_set v) :=
+set.fintype_subset (G''.neighbor_set v) (neighbor_set_subset_of_subgraph h v)
+
+instance coe_finite_at {G' : subgraph G} (v : G'.verts) [fintype (G'.neighbor_set v)] :
+  fintype (G'.coe.neighbor_set v) :=
+fintype.of_equiv _ (coe_neighbor_set_equiv v).symm
+
+/-- The degree of a vertex in a subgraph.  Is zero for vertices outside the subgraph. -/
+def degree (G' : subgraph G) (v : V) [fintype (G'.neighbor_set v)] : ℕ :=
```

Comment:
I'm still not super convinced about this. `neighbour_set` makes more sense to me, as it's so vital, but this seems to be one single `.coe ↑` away from the other, and in order to do anything useful you'd need to use `coe_degree` anyways.

Reply:
I don't agree with your assessment, but I'm not sure what to say since you don't give any examples or suggestions.

I'll just mention the intent with the design.  This lets you easily compare degrees of vertices of different subgraphs of the same graph without needing to wrangle proofs that the vertices are actually present in the subgraphs.  Take a look at lemmas like `degree_le'` -- think about how those would go if it was all in terms of `.coe.degree`.  Before switching to doing it this way, there used to be functions to coerce vertices of `G'.coe` to `G''.coe` using a `G' ≤ G''`, but it was all a pain to deal with.